src/openzeppelin/utils/presets/UniversalDeployer.cairo

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts for Cairo v0.4.0b (utils/presets/UniversalDeployer.cairo)
+
+%lang starknet
+
+from starkware.starknet.common.syscalls import get_caller_address, deploy
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.hash import hash2
+from starkware.cairo.common.bool import FALSE, TRUE
+
+@event
+func ContractDeployed(
+    address: felt,
+    deployer: felt,
+    unique: felt,
+    classHash: felt,
+    calldata_len: felt,
+    calldata: felt*,
+    salt: felt
+) {
+}
+
+@external
+func deployContract{
+    syscall_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    range_check_ptr
+}(
+    classHash: felt,
+    salt: felt,
+    unique: felt,
+    calldata_len: felt,
+    calldata: felt*
+) -> (address: felt) {
+    alloc_locals;
+    let (deployer) = get_caller_address();
+
+    tempvar prefix;
+    tempvar from_zero;
+    if (unique == TRUE) {
+        prefix = deployer;
+        from_zero = FALSE;
+    } else {
+        prefix = 'UniversalDeployerContract';
+        from_zero = TRUE;
+    }
+
+    let (_salt) = hash2{hash_ptr=pedersen_ptr}(prefix, salt);
+
+    let (local address) = deploy(
+        class_hash=classHash,
+        contract_address_salt=_salt,
+        constructor_calldata_size=calldata_len,
+        constructor_calldata=calldata,
+        deploy_from_zero=from_zero,
+    );
+
+    ContractDeployed.emit(
+        address=address,
+        deployer=deployer,
+        unique=unique,
+        classHash=classHash,
+        calldata_len=calldata_len,
+        calldata=calldata,
```

Comment:
@bbrandtom i'd like to confirm this is the right way of logging an array. Since `calldata` here is just the beginning of the array, [on the python tests](https://github.com/OpenZeppelin/cairo-contracts/pull/467/files#diff-22ef7551fe88a740023317f626698d866597930332e276b1b9946a61b4a78999R101) we need to compare it against `calldata[0]` for it to pass.

```python
        data=[
            deployed_address,               # contractAddress
            account.contract_address,  # deployer
            unique,                                   # unique
            class_hash,                            # classHash
            len(calldata),                         # calldata_len
            calldata[0],                            # calldata
            salt,                                        # salt
        ]
```

There's two things I'm not sure about:

* Is this enough to test the entire `calldata` is being logged? Or does the immutable memory model guarantee that the whole array will always be accessible from this pointer + traversing with `calldata_len`? I'd like this to catch e.g. if a buggy implementation is only logging the first item.
* I'm a bit confused since I would have expected python's `calldata[0]` to equal Cairo's `[calldata]` (dereference) instead of `calldata` (pointer).

Reply:
fixed it!